### ISSUE TITLE
Msgpack interoperability

### ISSUE THREAD

**Issue by ggtools** [ugorji/go#283]
## Versions

go: `go version go1.11.5 darwin/amd64`
codec: `github.com/ugorji/go/codec v0.0.0-20190204201341-e444a5086c43``

## Background

I'm working on a project which uses msgpack between Python, C# and now go. For various reasons we already had some *rules* which are:

1. Use the new byte type
2. Encode all timestamp as float seconds from Epoch
3. Encode UUID as strings

## Testing

I'm trying to find a set of parameters allowing those rules to be followed especially to convert automatically the UUID to strings. 

### Bytes

```go
var b = make([]byte, 0, 1024)
var h = new(codec.MsgpackHandle)
h.WriteExt = ???

var msg interface{}
var enc = codec.NewEncoderBytes(&b, h)
msg = []byte{0x82, 0x83}
enc.MustEncode(msg)
fmt.Printf("[]byte: % x\n", b)
```

With `h.WriteExt` set to  `false`, the byte array is encoded as `a2 82 83` which is a string so not compliant with rule 1. When set to `true` the  byte array is encoded as `c4 02 82 83` which is a 2 byte long binary object.

### UUID

I'm taking the previous code and in addition to replacing msg with an UUID, I added an extension following the [primer](http://ugorji.net/blog/go-codec-primer#using-extensions).

Extension code:

```go
// UUIDExt convert UUID to string
type UUIDExt struct{}

// WriteExt convert a uuid to a string
func (x UUIDExt) WriteExt(v interface{}) []byte {
fmt.Println("Converting")
switch v2 := v.(type) {
case uuid.UUID:
	return []byte(v2.String())
case *uuid.UUID:
	return []byte(v2.String())
default:
	panic(fmt.Sprintf("Unsupported type %T", v))
}

// ReadExt omitted
``` 

And in the main file I inserted:

```go
uuidTyp := reflect.TypeOf(uuid.UUID{})
var uuidExt UUIDExt
h.SetBytesExt(uuidTyp, 0xaa, uuidExt) // 0xaa is easy to spot in a binary stream
```

With `h.WriteExt` set to  `false`, the UUID is encoded as: `da 00 24 64 32 36 35 38 36 33 36 2d 61 39 36 65 2d 34 33 64 39 2d 38 66 31 35 2d 31 34 65 34 61 62 39 39 63 33 37 35` which is a string as expected. However when set to `true` the result is `c7 24 aa 64 32 36 35 38 36 33 36 2d 61 39 36 65 2d 34 33 64 39 2d 38 66 31 35 2d 31 34 65 34 61 62 39 39 63 33 37 35` which the same 0x24 byte payload **but** with an extension instead of a string so definitely not compliant with my rule 3.

### Time

TLDR: regardless of `WriteExt` and `TimeNotBuiltin` I can only get time encoded as the 64 bit either as byte array with `WriteExt` set to `false` or as the specific time extension.

## Problems

So I have 3 problems here:

1. The `WriteExt` flag seems to have a dual purpose as it triggers the use of binary type for byte arrays but also change the way extended types are managed. Unfortunately for me I have to wait for the quantum version of codec since I need the flag to be set to both `true` and `false`
2. With `WriteExt` set to `false`, the UUID is encoding using the `0xda` (str16) type while the length (0x24 bytes) would allow to use the `0xd9` (str8) type
3. The ext mechanism does not seem to work for the `time.Time` type.

**Comment by ugorji** [comment]
Solution

- Rule 1: WriteExt=true
- Rule 2: TimeNotBuiltin=true
- Rule 2, 3: custom types (e.g. UUID, custom time.Time wrapper) implement codec.Selfer or encoding.Binary(M|Unm)arshaler

Extensions by definition have an extension byte stored in the stream (per the format spec). If you just want to have custom encoding/decoding without extension byte, then it cannot be an extension - it must implement codec.Selfer of encoding.Binary(M|Unm)arshaler.

The problem you have, is that msgpack already defines an extension for time.Time. See https://github.com/msgpack/msgpack/blob/master/spec.md#timestamp-extension-type 

We support it natively now, and use the TimeNotBuiltin to say "disregard it". Then time.Time also implements encoding.Binary(M|Unm)arshaler, so that takes effect by default. Unless you define an extension (as you did) - and then we treat it as an extension. 

See https://godoc.org/github.com/ugorji/go/codec#hdr-Custom_Encoding_and_Decoding - where the algo is specified.

The way the library works is clear and flexible, but I think you should use the msgpack-defined time.Time extension, and then you will be all set. I think trying to force how time.Time is encoded/decoded is what is causing issues, as it is not a user-defined type, and the author already defined what it implements.

Hope this helps.
